src/box/lua/config/applier/box_cfg.lua

```diff
@@ -92,6 +93,108 @@ local function switch_isolated_mode(configdata, box_cfg)
+    local configdata = config._configdata
+
+    -- If the isolated mode is not enabled, there is nothing to do.
+    if not configdata:get('isolated', {use_default = true}) then
+        return
+    end
+
+    -- Drop existing iproto connections.
+    --
+    -- An instance in the isolated mode shouldn't process a
+    -- traffic from a user.
+    --
+    -- Here we assume that when box.cfg() is finished, iproto
+    -- listening sockets are already closed and no new connections
+    -- may be created. So, the only thing to do is to close the
+    -- existing connections.
+    --
+    -- NB: It seems safe to start this fiber several times in
+    -- parallel, so there are no checks, whether such a fiber
+    -- already exists.
+    fiber.new(function()
+        local name = 'isolated_mode_drop_iproto_connections'
+        fiber.self():name(name, {truncate = true})
+
+        -- A note regarding box.iproto.internal.drop_connections.
+        --
+        -- It seems, the current behavior if connections are not
+        -- closed in time is to report the timeout error and
+        -- continue closing connection from the given generation
+        -- in background.
+        --
+        -- It means that the passed timeout doesn't affect the
+        -- behavior of the process of closing connections itself
+        -- and can be only used to let the caller know whether
+        -- and when the connections (existed at the moment of the
+        -- invocation) are closed.
+        --
+        -- Taking it into account, the timeout in this code is
+        -- not configurable and chosen more or less arbitrarily.
+        -- It only affects whether and when an alert is issued.
+        --
+        -- If we find a need in a future, we can make it
+        -- configurable.
+        local timeout = 10
+
+        -- Testing requires to redefine the timeout. Use an error
+        -- injection for that.
+        local errinj_name = 'ERRINJ_CONFIG_DROP_CONNECTION_TIMEOUT'
+        local errinj_timeout = box.error.injection.get(errinj_name)
+        assert(type(errinj_timeout) == 'number')
+        if errinj_timeout >= 0 then
+            log.warn('isolated mode: %s redefines the iproto connections ' ..
+                'drop timeout: %d', errinj_name, errinj_timeout)
+            timeout = errinj_timeout
+        end
+
+        local start_time = clock.monotonic()
+        local ok, err = pcall(box.iproto.internal.drop_connections, timeout)
+        if not ok then
+            -- An unique key is set here to don't duplicate the
+            -- alert if the box_cfg.apply function is called twice
+            -- within the same startup/reconfiguration process.
+            --
+            -- Currently, it is possible on startup due to the
+            -- safe startup process. See c80b121544f6 ("config:
+            -- add safe startup mode") for details.
+            local key = 'isolated_mode_drop_iproto_connections_timeout'
```

Comment:
It is stated that this fiber could be started several times, and if several of them fail we will loose some of the errors, may be we could concatenate the error messages, or generate a unique key for each fiber?

Reply:
If the alerts have the same key, the last one replaces the previous one. So, we have one alert here at most. It seems meaningful, because they have the same message, wthout any additional (possibly different) details (such as a list of problematic connections).


However, it seems that I can discard the alert after a successful call. I'll do.

Reply:
Am I answered your question? Is the answer OK for you?

----

> However, it seems that I can discard the alert after a successful call. I'll do.

I've implemented it. I didn't add a test case, because it seems complicated to perform so fast reconfigurations to get such fibers actually executed in parallel and, moreover, one should go over the timeout, but another should succeed.

Anyway, `aboard:drop(key)` is just no-op if there is no alert. I think it is completely OK to just drop the alert if the action/condition related to it passed successfully.

```diff
diff --git a/src/box/lua/config/applier/box_cfg.lua b/src/box/lua/config/applier/box_cfg.lua
index 711d541295..ecf2724c4c 100644
--- a/src/box/lua/config/applier/box_cfg.lua
+++ b/src/box/lua/config/applier/box_cfg.lua
@@ -161,35 +161,51 @@ local function switch_isolated_mode_after_box_cfg(config)
 
         local start_time = clock.monotonic()
         local ok, err = pcall(box.iproto.internal.drop_connections, timeout)
-        if not ok then
-            -- An unique key is set here to don't duplicate the
-            -- alert if the box_cfg.apply function is called twice
-            -- within the same startup/reconfiguration process.
-            --
-            -- Currently, it is possible on startup due to the
-            -- safe startup process. See c80b121544f6 ("config:
-            -- add safe startup mode") for details.
-            local key = 'isolated_mode_drop_iproto_connections_timeout'
 
-            local message = ('isolated mode: can\'t drop iproto connections ' ..
+        -- An unique key is set here to don't duplicate the alert
+        -- if the box_cfg.apply function is called twice within
+        -- the same startup/reconfiguration process.
+        --
+        -- Currently, it is possible on startup due to the safe
+        -- startup process. See c80b121544f6 ("config: add safe
+        -- startup mode") for details.
+        --
+        -- Also, this key is used to drop the alert (if any) after
+        -- a successful `drop_connections` call.
+        local alert_key = 'isolated_mode_drop_iproto_connections_timeout'
+
+        if not ok then
+           local message = ('isolated mode: can\'t drop iproto connections ' ..
                 'during %d seconds (continued in background): %s'):format(
                 timeout, err)
 
             -- There is no clear point, when to drop this alert.
-            -- So, don't bother about it much.
             --
             -- Once appeared, it exists till the next
             -- reconfiguration, because all the alerts are dropped
             -- before applying the new configuration.
+            --
+            -- Also, dropped on a successful attempt to drop
+            -- connections.
             config._aboard:set({
                 type = 'warn',
                 message = message,
             }, {
-                key = key,
+                key = alert_key,
             })
             return
         end
 
+        -- If it appears that there were several reconfigurations
+        -- in row, each schedules a fiber to drop the connections,
+        -- but the fibers are executed after the reconfigurations,
+        -- and the first issues an alert, but the next
+        -- successfully dropped the connections, let's drop this
+        -- alert.
+        --
+        -- No-op if there is no alert with the given key.
+        config._aboard:drop(alert_key)
+
         local duration = clock.monotonic() - start_time
         log.info('isolated mode: successfully dropped iproto connections ' ..
             '(took %d seconds)', duration)
```

Reply:
Yeah, thanks for the explanation!